arch/rv32i/src/pmp.rs

```diff
@@ -490,210 +683,478 @@ impl<const MAX_AVAILABLE_REGIONS_OVER_TWO: usize> kernel::platform::mpu::MPU
 
         Ok(())
     }
 
     fn configure_mpu(&self, config: &Self::MpuConfig) {
-        // Is the PMP already configured for this app?
-        let last_configured_for_this_app = self
-            .last_configured_for
-            .map_or(false, |last_id| last_id == config.id);
-
-        // Skip PMP configuration if it is already configured for this app and the MPU
-        // configuration of this app has not changed.
-        if !last_configured_for_this_app || config.is_dirty.get() {
-            for (x, region) in config.regions.iter().enumerate() {
-                let disable_val = (csr::pmpconfig::pmpcfg::r0::CLEAR
-                    + csr::pmpconfig::pmpcfg::w0::CLEAR
-                    + csr::pmpconfig::pmpcfg::x0::CLEAR
-                    + csr::pmpconfig::pmpcfg::a0::OFF)
-                    .value;
-
-                let (region_shift, other_region_mask) = if x % 2 == 0 {
-                    (0, 0xFFFF_0000)
-                } else {
-                    (16, 0x0000_FFFF)
-                };
-
-                match region {
-                    Some(r) => {
-                        let cfg_val = r.cfg.value as usize;
-                        let start = r.location.0 as usize;
-                        let size = r.location.1;
-
-                        csr::CSR.pmpconfig_set(
-                            x / 2,
-                            (disable_val | cfg_val << 8) << region_shift
-                                | (csr::CSR.pmpconfig_get(x / 2) & other_region_mask),
-                        );
-                        csr::CSR.pmpaddr_set(x * 2, (start) >> 2);
-                        csr::CSR.pmpaddr_set((x * 2) + 1, (start + size) >> 2);
-                    }
-                    None => {
-                        // Invalidate other regions not used in this PMPConfig.
-                        csr::CSR.pmpconfig_set(
-                            x / 2,
-                            (disable_val | (disable_val << 8)) << region_shift
-                                | (csr::CSR.pmpconfig_get(x / 2) & other_region_mask),
-                        );
-                    }
-                };
-            }
-            config.is_dirty.set(false);
-            self.last_configured_for.replace(config.id);
+        if !self.last_configured_for.contains(&config.id) || config.is_dirty.get() {
+            self.pmp.configure_pmp(&config.regions).unwrap()
         }
     }
 }
 
-/// This is PMP support for kernel regions
-/// PMP does not allow a deny by default option, so all regions not marked
-/// with the below commands will have full access.
-/// This is still a useful implementation as it can be used to limit the
-/// kernels access, for example removing execute permission from regions
-/// we don't need to execute from and removing write permissions from
-/// executable regions.
-impl<const MAX_AVAILABLE_REGIONS_OVER_TWO: usize> kernel::platform::mpu::KernelMPU
-    for PMP<MAX_AVAILABLE_REGIONS_OVER_TWO>
-{
```

Comment:
This should continue to be supported

As https://github.com/tock/tock/issues/3599 points out, this is difficult to add support later. We really should support this in the original design and hopefully avoid the situation we are in now

Reply:
I really think that the trait-based structure proposed here would allow us to add an implementation for this later, but I acknowledge that me just making this claim is less convincing than proof in the form of an actual implementation. Let me whip up an implementation which still supports this (albeit less flexible than previously).

Reply:
I added the `KernelProtectionPMP`, which demonstrates that the abstractions proposed here are flexible enough to continue supporting this type of PMP implementation.